include/envoy/network/listener.h

```diff
@@ -114,6 +114,46 @@ class ListenerCallbacks {
   virtual void onNewConnection(ConnectionPtr&& new_connection) PURE;
 };
 
+/**
+ * UDP listener callbacks.
+ */
+class UdpListenerCallbacks {
+public:
+  enum class ErrorCode { SYSCALL_ERROR, UNKNOWN_ERROR };
+
+  virtual ~UdpListenerCallbacks() = default;
+
+  /**
+   * On first packet received on a UDP socket. This allows the callback handler
+   * to establish filter chain (or any other prepararion).
+   *
+   * @param local_address Local bound socket network address.
+   * @param peer_address Network address of the peer.
+   * @param data Data buffer received.
+   */
+  virtual void onNewConnection(Address::InstanceConstSharedPtr local_address,
```

Comment:
I assume for the purpose of this PR, we are looking at stateless vs. connection oriented proxy, right? I guess my question comes down to, at this layer, what does `onNewConnection()` give us. It seems like there must be internal tracking to figure out whether this is "new", but then how are connections ever going to get cleaned up? Where is the interface for that? It seems like at this datagram layer, we could just have `onData()` and these are the interfaces for raw datagram listening. If we want connection oriented, could we just use the normal ListenerCallbacks which spit out normal connections which are then UdpConnectionImpl? Thoughts on this?

Reply:
@mattklein123  
Thanks for reviewing.

What about layering the connection concept on top of UDP/datagram? This would allow us to have 1 common `buffer` layer and then connections (eg QUIC) on top of it. When a connection oriented filter (QUIC) gets the `onNewConnection` event, it can then decide to do things like 1) prepare the protocol layers 2) Carve a protocol specific `Packet` and send it up the protocol layer.

Reply:
The problem is I don't understand how this layer is going to do conn tracking if the API looks like this. I think you need an API to "delete" a connection if that's the case?

Reply:
But then I think the `onNewConnection` API needs to be removed? What does a new connection mean without conn tracking?

Reply:
Yes that is a good point. The book keeping of `if this is the first data packet` has to be done somewhere, assuming that the state of `first data` is important to track. If it is not, then the API `onNewConnection` need not be there.

Reply:
Yeah IMO if we are going to build connection oriented on top of this, I would kill `onNewConnection` and any tracking from this PR.